

<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta charset="utf-8" />
    <title>createSimpleISAjson.py &#8212; The ISA cookbook</title>
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.2/css/all.min.css" integrity="sha384-KA6wR/X5RY4zFAHpv/CnoG2UW1uogYfdnP67Uv7eULvTveboZJg0qUpmJZb5VqzN" crossorigin="anonymous">
    <link href="../../_static/css/index.css" rel="stylesheet">
    <link rel="stylesheet" href="../../_static/sphinx-book-theme.css" type="text/css" />
    <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/custom.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/jupyter-sphinx.css" />
    <script id="documentation_options" data-url_root="../../" src="../../_static/documentation_options.js"></script>
    <script src="../../_static/sphinx-book-theme.js"></script>
    <script src="../../_static/jquery.js"></script>
    <script src="../../_static/underscore.js"></script>
    <script src="../../_static/doctools.js"></script>
    <script src="../../_static/language_data.js"></script>
    <script src="../../_static/togglebutton.js"></script>
    <script src="../../_static/clipboard.min.js"></script>
    <script src="../../_static/copybutton.js"></script>
    <script src="../../_static/mystnb.js"></script>
    <script src="../../_static/mermaid.js"></script>
    <script src="../../_static/mermaid.min.js"></script>
    <script src="../../_static/mermaid.core.js"></script>
    <script src="../../_static/sphinx-book-theme.js"></script>
    <script >var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="https://unpkg.com/@jupyter-widgets/html-manager@^0.18.0/dist/embed-amd.js"></script>
    <link rel="canonical" href="https://isa-tools.github.io/isa-cookbook/content/notebooks/example-createSimpleISAJSON.html" />
    <link rel="shortcut icon" href="../../_static/isa.ico"/>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />

    <meta name="viewport" content="width=device-width, initial-scale=1">
    <meta name="docsearch:language" content="en">


<!-- Opengraph tags -->
<meta property="og:url"         content="https://isa-tools.github.io/isa-cookbook/content/notebooks/example-createSimpleISAJSON.html" />
<meta property="og:type"        content="article" />
<meta property="og:title"       content="createSimpleISAjson.py" />
<meta property="og:description" content="createSimpleISAjson.py  An example of using the ISA model classes to create an ISA-JSON file.  #!/usr/bin/env python  from isatools.model import *   def create_" />
<meta property="og:image"       content="https://isa-tools.github.io/isa-cookbook/_static/isa_footer_logo.svg" />

<meta name="twitter:card" content="summary">


  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="80">
    

    <div class="container-xl">
      <div class="row">
          
<div class="col-12 col-md-3 bd-sidebar site-navigation show" id="site-navigation">
    
        <div class="navbar-brand-box">
<a class="navbar-brand text-wrap" href="../../index.html">
  
  <img src="../../_static/isa_footer_logo.svg" class="logo" alt="logo">
  
  
  <h1 class="site-logo" id="site-title">The ISA cookbook</h1>
  
</a>
</div>

<form class="bd-search d-flex align-items-center" action="../../search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form>

<nav class="bd-links" id="bd-docs-nav" aria-label="Main navigation">
  
  <ul class="nav sidenav_l1">
  <li class="">
    <a href="../index.html">1. ISA tools API</a>
  </li>
  <li class="">
    <a href="../installation.html">2. ISA-API Installation</a>
  </li>
  <li class="">
    <a href="../isamodel.html">3. ISA data model</a>
  </li>
  <li class="">
    <a href="../reading.html">4. Reading in ISA-Tab or ISA JSON</a>
  </li>
  <li class="">
    <a href="../creation.html">5. Creating ISA content</a>
  </li>
  <li class="">
    <a href="../validation.html">6. Validating ISA-Tab and ISA JSON</a>
  </li>
  <li class="">
    <a href="../conversions.html">7. Converting between ISA format</a>
  </li>
  <li class="">
    <a href="../importdata.html">8. Importing data into ISA formats</a>
  </li>
  <li class="">
    <a href="../exportdata.html">9. Exporting data from ISA formats</a>
  </li>
  <li class="">
    <a href="../examples.html">10. Example programs using the ISA API</a>
  </li>
  <li class="">
    <a href="../github.html">11. Downloading files stored in Github</a>
  </li>
  <li class="">
    <a href="../knownissues.html">12. Known issues</a>
  </li>
  <li class="">
    <a href="isa-create-sample-assay-plan-htp-imaging.html">13. ISA Create - Sample Assay Plan as a Graph: Phenotyping by HT imaging Spectrometry</a>
  </li>
  <li class="">
    <a href="isa-create-sample-assay-plan-lc-dad.html">14. ISA Create - Sample Assay Plan as a Graph: LC-Diode-Array-Assay Spectrometry</a>
  </li>
</ul>
</nav>

 <!-- To handle the deprecated key -->

<div class="navbar_extra_footer">
  Powered by <a href="https://jupyterbook.org">Jupyter Book</a>
</div>

</div>


          


          
<main class="col py-md-3 pl-md-4 bd-content overflow-auto" role="main">
    
    <div class="row topbar fixed-top container-xl">
    <div class="col-12 col-md-3 bd-topbar-whitespace site-navigation show">
    </div>
    <div class="col pl-2 topbar-main">
        
        <button id="navbar-toggler" class="navbar-toggler ml-0" type="button" data-toggle="collapse"
            data-toggle="tooltip" data-placement="bottom" data-target=".site-navigation" aria-controls="navbar-menu"
            aria-expanded="true" aria-label="Toggle navigation" aria-controls="site-navigation"
            title="Toggle navigation" data-toggle="tooltip" data-placement="left">
            <i class="fas fa-bars"></i>
            <i class="fas fa-arrow-left"></i>
            <i class="fas fa-arrow-up"></i>
        </button>
        
        <div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn" aria-label="Download this page"><i
            class="fas fa-download"></i></button>

    
    <div class="dropdown-buttons">
        <!-- ipynb file if we had a myst markdown file -->
        
        <!-- Download raw file -->
        <a class="dropdown-buttons" href="../../_sources/content/notebooks/example-createSimpleISAJSON.md"><button type="button"
                class="btn btn-secondary topbarbtn" title="Download source file" data-toggle="tooltip"
                data-placement="left">.md</button></a>
        <!-- Download PDF via print -->
        <button type="button" id="download-print" class="btn btn-secondary topbarbtn" title="Print to PDF"
            onClick="window.print()" data-toggle="tooltip" data-placement="left">.pdf</button>
    </div>
    
</div>
        <!-- Source interaction buttons -->

<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn"
        aria-label="Connect with source repository"><i class="fab fa-github"></i></button>
    <div class="dropdown-buttons sourcebuttons">
        <a class="repository-button"
            href="https://github.com/isa-tools/isa-cookbook"><button type="button" class="btn btn-secondary topbarbtn"
                data-toggle="tooltip" data-placement="left" title="Source repository"><i
                    class="fab fa-github"></i>repository</button></a>
        <a class="issues-button"
            href="https://github.com/isa-tools/isa-cookbook/issues/new?title=Issue%20on%20page%20%2Fcontent/notebooks/example-createSimpleISAJSON.html&body=Your%20issue%20content%20here."><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Open an issue"><i class="fas fa-lightbulb"></i>open issue</button></a>
        <a class="edit-button" href="https://github.com/isa-tools/isa-cookbook/edit/main/isa-book/content/notebooks/example-createSimpleISAJSON.md"><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Edit this page"><i class="fas fa-pencil-alt"></i>suggest edit</button></a>
    </div>
</div>


        <!-- Full screen (wrap in <a> to have style consistency -->
        <a class="full-screen-button"><button type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip"
                data-placement="bottom" onclick="toggleFullScreen()" title="Fullscreen mode"><i
                    class="fas fa-expand"></i></button></a>

        <!-- Launch buttons -->

    </div>
    <div class="d-none d-md-block col-md-2 bd-toc show">

<nav id="bd-toc-nav">
    <ul class="nav section-nav flex-column">
    
    </ul>
</nav>



<div class="tocsection editthispage">
    <a href="https://github.com/isa-tools/isa-cookbook/edit/main/isa-book/content/notebooks/example-createSimpleISAJSON.md">
        <i class="fas fa-pencil-alt"></i> Edit this page
    </a>
</div>

    </div>
</div>
    <div id="main-content" class="row">
        <div class="col-12 col-md-9 pl-md-3 pr-md-0">
        
              <div>
                
  <div class="section" id="createsimpleisajson-py">
<h1>createSimpleISAjson.py<a class="headerlink" href="#createsimpleisajson-py" title="Permalink to this headline">¶</a></h1>
<p>An example of using the ISA model classes to create an ISA-JSON file.</p>
<div class="highlight-{.sourceCode notranslate"><div class="highlight"><pre><span></span>#!/usr/bin/env python

from isatools.model import *


def create_descriptor():
    &quot;&quot;&quot;Returns a simple but complete ISA-JSON 1.0 descriptor for illustration.&quot;&quot;&quot;

    # Create an empty Investigation object and set some values to the instance variables.

    investigation = Investigation()
    investigation.identifier = &quot;i1&quot;
    investigation.title = &quot;My Simple ISA Investigation&quot;
    investigation.description = &quot;We could alternatively use the class constructor&#39;s parameters to set some default &quot; \
                                &quot;values at the time of creation, however we want to demonstrate how to use the &quot; \
                                &quot;object&#39;s instance variables to set values.&quot;
    investigation.submission_date = &quot;2016-11-03&quot;
    investigation.public_release_date = &quot;2016-11-03&quot;

    # Create an empty Study object and set some values. The Study must have a filename, otherwise when we serialize it
    # to ISA-Tab we would not know where to write it. We must also attach the study to the investigation by adding it
    # to the &#39;investigation&#39; object&#39;s list of studies.

    study = Study(filename=&quot;s_study.txt&quot;)
    study.identifier = &quot;s1&quot;
    study.title = &quot;My ISA Study&quot;
    study.description = &quot;Like with the Investigation, we could use the class constructor to set some default values, &quot; \
                        &quot;but have chosen to demonstrate in this example the use of instance variables to set initial &quot; \
                        &quot;values.&quot;
    study.submission_date = &quot;2016-11-03&quot;
    study.public_release_date = &quot;2016-11-03&quot;
    investigation.studies.append(study)

    # Some instance variables are typed with different objects and lists of objects. For example, a Study can have a
    # list of design descriptors. A design descriptor is an Ontology Annotation describing the kind of study at hand.
    # Ontology Annotations should typically reference an Ontology Source. We demonstrate a mix of using the class
    # constructors and setting values with instance variables. Note that the OntologyAnnotation object
    # &#39;intervention_design&#39; links its &#39;term_source&#39; directly to the &#39;obi&#39; object instance. To ensure the OntologySource
    # is encapsulated in the descriptor, it is added to a list of &#39;ontology_source_references&#39; in the Investigation
    # object. The &#39;intervention_design&#39; object is then added to the list of &#39;design_descriptors&#39; held by the Study
    # object.

    obi = OntologySource(name=&#39;OBI&#39;, description=&quot;Ontology for Biomedical Investigations&quot;)
    investigation.ontology_source_references.append(obi)
    intervention_design = OntologyAnnotation(term_source=obi)
    intervention_design.term = &quot;intervention design&quot;
    intervention_design.term_accession = &quot;http://purl.obolibrary.org/obo/OBI_0000115&quot;
    study.design_descriptors.append(intervention_design)

    # Other instance variables common to both Investigation and Study objects include &#39;contacts&#39; and &#39;publications&#39;,
    # each with lists of corresponding Person and Publication objects.

    contact = Person(first_name=&quot;Alice&quot;, last_name=&quot;Robertson&quot;, affiliation=&quot;University of Life&quot;, roles=[OntologyAnnotation(term=&#39;submitter&#39;)])
    study.contacts.append(contact)
    publication = Publication(title=&quot;Experiments with Elephants&quot;, author_list=&quot;A. Robertson, B. Robertson&quot;)
    publication.pubmed_id = &quot;12345678&quot;
    publication.status = OntologyAnnotation(term=&quot;published&quot;)
    study.publications.append(publication)

    # To create the study graph that corresponds to the contents of the study table file (the s_*.txt file), we need
    # to create a process sequence. To do this we use the Process class and attach it to the Study object&#39;s
    # &#39;process_sequence&#39; list instance variable. Each process must be linked with a Protocol object that is attached to
    # a Study object&#39;s &#39;protocols&#39; list instance variable. The sample collection Process object usually has as input
    # a Source material and as output a Sample material.

    # Here we create one Source material object and attach it to our study.

    source = Source(name=&#39;source_material&#39;)
    study.sources.append(source)

    # Then we create three Sample objects, with organism as Homo Sapiens, and attach them to the study. We use the utility function
    # batch_create_material() to clone a prototype material object. The function automatically appends
    # an index to the material name. In this case, three samples will be created, with the names
    # &#39;sample_material-0&#39;, &#39;sample_material-1&#39; and &#39;sample_material-2&#39;.

    prototype_sample = Sample(name=&#39;sample_material&#39;, derives_from=[source])
    ncbitaxon = OntologySource(name=&#39;NCBITaxon&#39;, description=&quot;NCBI Taxonomy&quot;)
    characteristic_organism = Characteristic(category=OntologyAnnotation(term=&quot;Organism&quot;),
                                     value=OntologyAnnotation(term=&quot;Homo Sapiens&quot;, term_source=ncbitaxon,
                                                              term_accession=&quot;http://purl.bioontology.org/ontology/NCBITAXON/9606&quot;))
    prototype_sample.characteristics.append(characteristic_organism)

    study.samples = batch_create_materials(prototype_sample, n=3)  # creates a batch of 3 samples

    # Now we create a single Protocol object that represents our sample collection protocol, and attach it to the
    # study object. Protocols must be declared before we describe Processes, as a processing event of some sort
    # must execute some defined protocol. In the case of the class model, Protocols should therefore be declared
    # before Processes in order for the Process to be linked to one.

    sample_collection_protocol = Protocol(name=&quot;sample collection&quot;,
                                          protocol_type=OntologyAnnotation(term=&quot;sample collection&quot;))
    study.protocols.append(sample_collection_protocol)
    sample_collection_process = Process(executes_protocol=sample_collection_protocol)

    # Next, we link our materials to the Process. In this particular case, we are describing a sample collection
    # process that takes one source material, and produces three different samples.
    #
    # (source_material)-&gt;(sample collection)-&gt;[(sample_material-0), (sample_material-1), (sample_material-2)]

    for src in study.sources:
        sample_collection_process.inputs.append(src)
    for sam in study.samples:
        sample_collection_process.outputs.append(sam)

    # Finally, attach the finished Process object to the study process_sequence. This can be done many times to
    # describe multiple sample collection events.

    study.process_sequence.append(sample_collection_process)

    # Next, we build n Assay object and attach two protocols, extraction and sequencing.

    assay = Assay(filename=&quot;a_assay.txt&quot;)
    extraction_protocol = Protocol(name=&#39;extraction&#39;, protocol_type=OntologyAnnotation(term=&quot;material extraction&quot;))
    study.protocols.append(extraction_protocol)
    sequencing_protocol = Protocol(name=&#39;sequencing&#39;, protocol_type=OntologyAnnotation(term=&quot;material sequencing&quot;))
    study.protocols.append(sequencing_protocol)

    # To build out assay graphs, we enumereate the samples from the study-level, and for each sample we create an
    # extraction process and a sequencing process. The extraction process takes as input a sample material, and produces
    # an extract material. The sequencing process takes the extract material and produces a data file. This will
    # produce three graphs, from sample material through to data, as follows:
    #
    # (sample_material-0)-&gt;(extraction)-&gt;(extract-0)-&gt;(sequencing)-&gt;(sequenced-data-0)
    # (sample_material-1)-&gt;(extraction)-&gt;(extract-1)-&gt;(sequencing)-&gt;(sequenced-data-1)
    # (sample_material-2)-&gt;(extraction)-&gt;(extract-2)-&gt;(sequencing)-&gt;(sequenced-data-2)
    #
    # Note that the extraction processes and sequencing processes are distinctly separate instances, where the three
    # graphs are NOT interconnected.

    for i, sample in enumerate(study.samples):

        # create an extraction process that executes the extraction protocol

        extraction_process = Process(executes_protocol=extraction_protocol)

        # extraction process takes as input a sample, and produces an extract material as output

        extraction_process.inputs.append(sample)
        material = Material(name=&quot;extract-{}&quot;.format(i))
        material.type = &quot;Extract Name&quot;
        extraction_process.outputs.append(material)

        # create a sequencing process that executes the sequencing protocol

        sequencing_process = Process(executes_protocol=sequencing_protocol)
        sequencing_process.name = &quot;assay-name-{}&quot;.format(i)
        sequencing_process.inputs.append(extraction_process.outputs[0])

        # Sequencing process usually has an output data file

        datafile = DataFile(filename=&quot;sequenced-data-{}&quot;.format(i), label=&quot;Raw Data File&quot;, generated_from=[sample])
        sequencing_process.outputs.append(datafile)

        # Ensure Processes are linked forward and backward. plink(from_process, to_process) is a function to set
        # these links for you. It is found in the isatools.model package

        plink(extraction_process, sequencing_process)

        # make sure the extract, data file, and the processes are attached to the assay

        assay.samples.append(sample)
        assay.data_files.append(datafile)
        assay.other_material.append(material)
        assay.process_sequence.append(extraction_process)
        assay.process_sequence.append(sequencing_process)
        assay.measurement_type = OntologyAnnotation(term=&quot;gene sequencing&quot;)
        assay.technology_type = OntologyAnnotation(term=&quot;nucleotide sequencing&quot;)

    # attach the assay to the study

    study.assays.append(assay)

    import json
    from isatools.isajson import ISAJSONEncoder

    # To write JSON out, use the ISAJSONEncoder class with the json package and use dump() or dumps()
    # Note that the extra parameters sort_keys, indent and separators are to make the output more human-readable.

    return json.dumps(investigation, cls=ISAJSONEncoder, sort_keys=True, indent=4, separators=(&#39;,&#39;, &#39;: &#39;))

if __name__ == &#39;__main__&#39;:
    print(create_descriptor())  # print the result to stdout
</pre></div>
</div>
</div>


              </div>
              
        </div>
    </div>
    
    
    <div class='prev-next-bottom'>
        

    </div>
    <footer class="footer mt-5 mt-md-0">
    <div class="container">
      <p>
        
          By The ISA Team<br/>
        
            &copy; Copyright 2021 (CC-BY 4.0).<br/>
          <div class="extra_footer">
            <script>mermaid.init();</script> <div class="row_footer"> This page was created thanks to work by <a href="https://github.com/jupyter/jupyter-book/graphs/contributors">The Jupyter Book Community</a></div> <div class="row_footer"> This project is supported financially by the <a href="https://wellcome.org/">Welcome Trust</a>, Elixir-UK and the European Union</div>
<div class="row_footer"> <div class="column" style="center"><a href="https://wellcome.org/"><img src="../_static/images/logo/wellcome-logo-black.png" height="25" alt="Wellcome" style="float:center""></a></div> <div class="column"><a href="https://www.elixir-uk.org.eu/"><img src="../_static/images/logo/elixir-uk_transparent.png" height="40" alt="IMI logo" style="float:center""></a></div> <div class="column"><a href="https://https://ec.europa.eu/programmes/horizon2020/en"><img src="../_static/images/logo/eu-flag-recolored.png" height="25" alt="EFPIA logo" style="float:center""></a></div> </div>
          </div>
      </p>
    </div>
  </footer>
</main>


      </div>
    </div>

    <script src="../../_static/js/index.js"></script>
    
  </body>
</html>