{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 146,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import isatools\n",
    "from isatools import isatab"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 147,
   "metadata": {},
   "outputs": [],
   "source": [
    "from isatools.model import (\n",
    "    Investigation,\n",
    "    Study,\n",
    "    OntologySource,\n",
    "    OntologyAnnotation,\n",
    "    Person,\n",
    "    Publication,\n",
    "    Source,\n",
    "    Sample,\n",
    "    Characteristic,\n",
    "    Protocol,\n",
    "    Process,\n",
    "    Assay,\n",
    "    Material,\n",
    "    DataFile,\n",
    "    batch_create_materials,\n",
    "    plink\n",
    ")\n",
    "import json\n",
    "from isatools.isajson import ISAJSONEncoder\n",
    "from isatools.isatab import dump_tables_to_dataframes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 148,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create an empty Investigation object and set some values to the instance variables.\n",
    "\n",
    "investigation = Investigation()\n",
    "# investigation.identifier = \"i1\"\n",
    "# investigation.title = \"My Simple ISA Investigation\"\n",
    "# investigation.description = \"We could alternatively use the class constructor's parameters to set some default \" \\\n",
    "#                             \"values at the time of creation, however we want to demonstrate how to use the \" \\\n",
    "#                             \"object's instance variables to set values.\"\n",
    "# investigation.submission_date = \"2016-11-03\"\n",
    "# investigation.public_release_date = \"201611-03\"\n",
    "\n",
    "# Create an empty Study object and set some values. The Study must have a filename, otherwise when we serialize it\n",
    "# to ISA-Tab we would not know where to write it. We must also attach the study to the investigation by adding it\n",
    "# to the 'investigation' object's list of studies.\n",
    "\n",
    "study = Study(filename=\"s_study.txt\")\n",
    "study.identifier = \"s1\"\n",
    "study.title = \"My ISA Study\"\n",
    "study.description = \"Like with the Investigation, we could use the class constructor to set some default values, \" \\\n",
    "                    \"but have chosen to demonstrate in this example the use of instance variables to set initial \" \\\n",
    "                    \"values.\"\n",
    "study.submission_date = \"2016-11-03\"\n",
    "study.public_release_date = \"2016-11-03\"\n",
    "investigation.studies.append(study)\n",
    "\n",
    "# Some instance variables are typed with different objects and lists of objects. For example, a Study can have a\n",
    "# list of design descriptors. A design descriptor is an Ontology Annotation describing the kind of study at hand.\n",
    "# Ontology Annotations should typically reference an Ontology Source. We demonstrate a mix of using the class\n",
    "# constructors and setting values with instance variables. Note that the OntologyAnnotation object\n",
    "# 'intervention_design' links its 'term_source' directly to the 'obi' object instance. To ensure the OntologySource\n",
    "# is encapsulated in the descriptor, it is added to a list of 'ontology_source_references' in the Investigation\n",
    "# object. The 'intervention_design' object is then added to the list of 'design_descriptors' held by the Study\n",
    "# object.\n",
    "\n",
    "obi = OntologySource(name='OBI', description=\"Ontology for Biomedical Investigations\")\n",
    "investigation.ontology_source_references.append(obi)\n",
    "intervention_design = OntologyAnnotation(term_source=obi)\n",
    "intervention_design.term = \"intervention design\"\n",
    "intervention_design.term_accession = \"http://purl.obolibrary.org/obo/OBI_0000115\"\n",
    "study.design_descriptors.append(intervention_design)\n",
    "\n",
    "# Other instance variables common to both Investigation and Study objects include 'contacts' and 'publications',\n",
    "# each with lists of corresponding Person and Publication objects.\n",
    "\n",
    "contact = Person(first_name=\"Alice\", last_name=\"Robertson\", affiliation=\"University of Life\", roles=[OntologyAnnotation(term='submitter')])\n",
    "study.contacts.append(contact)\n",
    "publication = Publication(title=\"Experiments with Elephants\", author_list=\"A. Robertson, B. Robertson\")\n",
    "publication.pubmed_id = \"12345678\"\n",
    "publication.doi = \"10.1038/sdata.2016.18\" #https://doi.org/10.144534/rmh0000008\"\n",
    "publication.status = OntologyAnnotation(term=\"published\")\n",
    "study.publications.append(publication)\n",
    "\n",
    "# To create the study graph that corresponds to the contents of the study table file (the s_*.txt file), we need\n",
    "# to create a process sequence. To do this we use the Process class and attach it to the Study object's\n",
    "# 'process_sequence' list instance variable. Each process must be linked with a Protocol object that is attached to\n",
    "# a Study object's 'protocols' list instance variable. The sample collection Process object usually has as input\n",
    "# a Source material and as output a Sample material.\n",
    "\n",
    "# Here we create one Source material object and attach it to our study.\n",
    "\n",
    "source = Source(name='source_material')\n",
    "study.sources.append(source)\n",
    "\n",
    "# Then we create three Sample objects, with organism as Homo Sapiens, and attach them to the study. We use the utility function\n",
    "# batch_create_material() to clone a prototype material object. The function automatiaclly appends\n",
    "# an index to the material name. In this case, three samples will be created, with the names\n",
    "# 'sample_material-0', 'sample_material-1' and 'sample_material-2'.\n",
    "\n",
    "prototype_sample = Sample(name='sample_material', derives_from=(source,))\n",
    "ncbitaxon = OntologySource(name='NCBITaxon', description=\"NCBI Taxonomy\")\n",
    "characteristic_organism = Characteristic(category=OntologyAnnotation(term=\"Organism\"),\n",
    "                                 value=OntologyAnnotation(term=\"Homo Sapiens\", term_source=ncbitaxon,\n",
    "                                                          term_accession=\"http://purl.bioontology.org/ontology/NCBITAXON/9606\"))\n",
    "prototype_sample.characteristics.append(characteristic_organism)\n",
    "\n",
    "study.samples = batch_create_materials(prototype_sample, n=3)  # creates a batch of 3 samples\n",
    "\n",
    "# Now we create a single Protocol object that represents our sample collection protocol, and attach it to the\n",
    "# study object. Protocols must be declared before we describe Processes, as a processing event of some sort\n",
    "# must execute some defined protocol. In the case of the class model, Protocols should therefore be declared\n",
    "# before Processes in order for the Process to be linked to one.\n",
    "\n",
    "sample_collection_protocol = Protocol(name=\"sample collection\",\n",
    "                                      protocol_type=OntologyAnnotation(term=\"sample collection\"))\n",
    "study.protocols.append(sample_collection_protocol)\n",
    "sample_collection_process = Process(executes_protocol=sample_collection_protocol)\n",
    "\n",
    "# Next, we link our materials to the Process. In this particular case, we are describing a sample collection\n",
    "# process that takes one source material, and produces three different samples.\n",
    "#\n",
    "# (source_material)->(sample collection)->[(sample_material-0), (sample_material-1), (sample_material-2)]\n",
    "\n",
    "for src in study.sources:\n",
    "    sample_collection_process.inputs.append(src)\n",
    "for sam in study.samples:\n",
    "    sample_collection_process.outputs.append(sam)\n",
    "\n",
    "# Finally, attach the finished Process object to the study process_sequence. This can be done many times to\n",
    "# describe multiple sample collection events.\n",
    "\n",
    "study.process_sequence.append(sample_collection_process)\n",
    "\n",
    "# Next, we build n Assay object and attach two protocols, extraction and sequencing.\n",
    "\n",
    "# assay = Assay(filename=\"a_assay.txt\")\n",
    "# extraction_protocol = Protocol(name='extraction', protocol_type=OntologyAnnotation(term=\"material extraction\"))\n",
    "# study.protocols.append(extraction_protocol)\n",
    "# sequencing_protocol = Protocol(name='sequencing', protocol_type=OntologyAnnotation(term=\"material sequencing\"))\n",
    "# study.protocols.append(sequencing_protocol)\n",
    "\n",
    "# # To build out assay graphs, we enumereate the samples from the study-level, and for each sample we create an\n",
    "# # extraction process and a sequencing process. The extraction process takes as input a sample material, and produces\n",
    "# # an extract material. The sequencing process takes the extract material and produces a data file. This will\n",
    "# # produce three graphs, from sample material through to data, as follows:\n",
    "# #\n",
    "# # (sample_material-0)->(extraction)->(extract-0)->(sequencing)->(sequenced-data-0)\n",
    "# # (sample_material-1)->(extraction)->(extract-1)->(sequencing)->(sequenced-data-1)\n",
    "# # (sample_material-2)->(extraction)->(extract-2)->(sequencing)->(sequenced-data-2)\n",
    "# #\n",
    "# # Note that the extraction processes and sequencing processes are distinctly separate instances, where the three\n",
    "# # graphs are NOT interconnected.\n",
    "\n",
    "# for i, sample in enumerate(study.samples):\n",
    "\n",
    "#     # create an extraction process that executes the extraction protocol\n",
    "\n",
    "#     extraction_process = Process(executes_protocol=extraction_protocol)\n",
    "\n",
    "#     # extraction process takes as input a sample, and produces an extract material as output\n",
    "\n",
    "#     extraction_process.inputs.append(sample)\n",
    "#     material = Material(name=\"extract-{}\".format(i))\n",
    "#     material.type = \"Extract Name\"\n",
    "#     extraction_process.outputs.append(material)\n",
    "\n",
    "#     # create a sequencing process that executes the sequencing protocol\n",
    "\n",
    "#     sequencing_process = Process(executes_protocol=sequencing_protocol)\n",
    "#     sequencing_process.name = \"assay-name-{}\".format(i)\n",
    "#     sequencing_process.inputs.append(extraction_process.outputs[0])\n",
    "\n",
    "#     # Sequencing process usually has an output data file\n",
    "\n",
    "#     datafile = DataFile(filename=\"sequenced-data-{}\".format(i), label=\"Raw Data File\")\n",
    "#     sequencing_process.outputs.append(datafile)\n",
    "\n",
    "#     # Ensure Processes are linked forward and backward. plink(from_process, to_process) is a function to set\n",
    "#     # these links for you. It is found in the isatools.model package\n",
    "\n",
    "#     plink(extraction_process, sequencing_process)\n",
    "\n",
    "#     # make sure the extract, data file, and the processes are attached to the assay\n",
    "\n",
    "#     assay.data_files.append(datafile)\n",
    "#     assay.samples.append(sample)\n",
    "#     assay.other_material.append(material)\n",
    "#     assay.process_sequence.append(extraction_process)\n",
    "#     assay.process_sequence.append(sequencing_process)\n",
    "#     assay.measurement_type = OntologyAnnotation(term=\"gene sequencing\")\n",
    "#     assay.technology_type = OntologyAnnotation(term=\"nucleotide sequencing\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Next, we build n Assay object and attach two protocols, extraction and sequencing.\n",
    "\n",
    "assay_1 = Assay(filename=\"a_assay_1.txt\")\n",
    "extraction_protocol = Protocol(name='extraction', protocol_type=OntologyAnnotation(term=\"material extraction\"))\n",
    "study.protocols.append(extraction_protocol)\n",
    "sequencing_protocol_1 = Protocol(name='sequencing_1', protocol_type=OntologyAnnotation(term=\"nucleic acid sequencing\"))\n",
    "study.protocols.append(sequencing_protocol_1)\n",
    "\n",
    "# To build out assay graphs, we enumereate the samples from the study-level, and for each sample we create an\n",
    "# extraction process and a sequencing process. The extraction process takes as input a sample material, and produces\n",
    "# an extract material. The sequencing process takes the extract material and produces a data file. This will\n",
    "# produce three graphs, from sample material through to data, as follows:\n",
    "#\n",
    "# (sample_material-0)->(extraction)->(extract-0)->(sequencing)->(sequenced-data-0)\n",
    "# (sample_material-1)->(extraction)->(extract-1)->(sequencing)->(sequenced-data-1)\n",
    "# (sample_material-2)->(extraction)->(extract-2)->(sequencing)->(sequenced-data-2)\n",
    "#\n",
    "# Note that the extraction processes and sequencing processes are distinctly separate instances, where the three\n",
    "# graphs are NOT interconnected.\n",
    "\n",
    "for i, sample in enumerate(study.samples):\n",
    "\n",
    "    # create an extraction process that executes the extraction protocol\n",
    "\n",
    "    extraction_process = Process(executes_protocol=extraction_protocol)\n",
    "\n",
    "    # extraction process takes as input a sample, and produces an extract material as output\n",
    "\n",
    "    extraction_process.inputs.append(sample)\n",
    "    material = Material(name=\"extract-{}\".format(i))\n",
    "    material.type = \"Extract Name\"\n",
    "    extraction_process.outputs.append(material)\n",
    "\n",
    "    # create a sequencing process that executes the sequencing protocol\n",
    "\n",
    "    sequencing_process = Process(executes_protocol=sequencing_protocol_1)\n",
    "    sequencing_process.name = \"assay-name-{}\".format(i)\n",
    "    sequencing_process.inputs.append(extraction_process.outputs[0])\n",
    "\n",
    "    # Sequencing process usually has an output data file\n",
    "\n",
    "    datafile = DataFile(filename=\"sequenced-data-{}\".format(i), label=\"Raw Data File\")\n",
    "    sequencing_process.outputs.append(datafile)\n",
    "\n",
    "    # Ensure Processes are linked forward and backward. plink(from_process, to_process) is a function to set\n",
    "    # these links for you. It is found in the isatools.model package\n",
    "\n",
    "    plink(extraction_process, sequencing_process)\n",
    "\n",
    "    # make sure the extract, data file, and the processes are attached to the assay\n",
    "\n",
    "    assay_1.data_files.append(datafile)\n",
    "    assay_1.samples.append(sample)\n",
    "    assay_1.other_material.append(material)\n",
    "    assay_1.process_sequence.append(extraction_process)\n",
    "    assay_1.process_sequence.append(sequencing_process)\n",
    "    assay_1.measurement_type = OntologyAnnotation(term=\"genome sequencing\"\n",
    "    assay_1.technology_type = OntologyAnnotation(term=\"nucleotide sequencing\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Next, we build n Assay object and attach two protocols, extraction and sequencing.\n",
    "\n",
    "assay_2 = Assay(filename=\"a_assay_2.txt\")\n",
    "extraction_protocol = Protocol(name='nucleic acid extraction', protocol_type=OntologyAnnotation(term=\"nucleic acid extraction\"))\n",
    "study.protocols.append(extraction_protocol)\n",
    "sequencing_protocol = Protocol(name='sequencing_2', protocol_type=OntologyAnnotation(term=\"nucleic acid sequencing\"))\n",
    "study.protocols.append(sequencing_protocol)\n",
    "dt_protocol = Protocol(name='analysis', protocol_type=OntologyAnnotation(term=\"sequence analysis data transformation\"))\n",
    "study.protocols.append(dt_protocol)\n",
    "\n",
    "# To build out assay graphs, we enumereate the samples from the study-level, and for each sample we create an\n",
    "# extraction process and a sequencing process. The extraction process takes as input a sample material, and produces\n",
    "# an extract material. The sequencing process takes the extract material and produces a data file. This will\n",
    "# produce three graphs, from sample material through to data, as follows:\n",
    "#\n",
    "# (sample_material-0)->(extraction)->(extract-0)->(sequencing)->(sequenced-data-0)\n",
    "# (sample_material-1)->(extraction)->(extract-1)->(sequencing)->(sequenced-data-1)\n",
    "# (sample_material-2)->(extraction)->(extract-2)->(sequencing)->(sequenced-data-2)\n",
    "#\n",
    "# Note that the extraction processes and sequencing processes are distinctly separate instances, where the three\n",
    "# graphs are NOT interconnected.\n",
    "\n",
    "for i, sample in enumerate(study.samples):\n",
    "\n",
    "    # create an extraction process that executes the extraction protocol\n",
    "\n",
    "    extraction_process = Process(executes_protocol=extraction_protocol)\n",
    "\n",
    "    # extraction process takes as input a sample, and produces an extract material as output\n",
    "\n",
    "    extraction_process.inputs.append(sample)\n",
    "    material = Material(name=\"extract-{}\".format(i))\n",
    "    material.type = \"Extract Name\"\n",
    "    extraction_process.outputs.append(material)\n",
    "\n",
    "    # create a sequencing process that executes the sequencing protocol\n",
    "\n",
    "    sequencing_process = Process(executes_protocol=sequencing_protocol)\n",
    "    sequencing_process.name = \"assay-name-{}\".format(i)\n",
    "    sequencing_process.inputs.append(extraction_process.outputs[0])\n",
    "\n",
    "    # Sequencing process usually has an output data file\n",
    "\n",
    "    datafile = DataFile(filename=\"sequenced-data-{}\".format(i), label=\"Raw Data File\")\n",
    "    sequencing_process.outputs.append(datafile)\n",
    "    \n",
    "    # Data Analysis process\n",
    "    dt_process = Process(executes_protocol=dt_protocol)\n",
    "    dt_process.name = \"dt-name-{}\".format(i)\n",
    "    dt_process.inputs.append(sequencing_process.outputs[0])\n",
    "    \n",
    "    derived_file = DataFile(filename=\"derived-data-{}\".format(i), label=\"Derived Data File\")\n",
    "    dt_process.outputs.append(derived_file)\n",
    "    \n",
    "    # Ensure Processes are linked forward and backward. plink(from_process, to_process) is a function to set\n",
    "    # these links for you. It is found in the isatools.model package\n",
    "\n",
    "    plink(extraction_process, sequencing_process)\n",
    "    plink(sequencing_process, dt_process)\n",
    "\n",
    "    # make sure the extract, data file, and the processes are attached to the assay\n",
    "\n",
    "    assay_2.data_files.append(datafile)\n",
    "    assay_2.samples.append(sample)\n",
    "    assay_2.other_material.append(material)\n",
    "    assay_2.process_sequence.append(extraction_process)\n",
    "    assay_2.process_sequence.append(sequencing_process)\n",
    "    assay_2.process_sequence.append(dt_process)\n",
    "    assay_2.measurement_type = OntologyAnnotation(term=\"transcription profiling\")\n",
    "    assay_2.technology_type = OntologyAnnotation(term=\"nucleotide sequencing\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 151,
   "metadata": {},
   "outputs": [],
   "source": [
    "# attach the assays to the study\n",
    "\n",
    "# study.assays.append(assay)\n",
    "study.assays.append(assay_1)\n",
    "study.assays.append(assay_2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 152,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2021-01-20 21:50:28,022 [INFO]: isatab.py(_all_end_to_end_paths:1152) >> Found 3 paths!\n",
      "2021-01-20 21:50:28,026 [INFO]: isatab.py(write_study_table_files:1330) >> Rendered 3 paths\n",
      "2021-01-20 21:50:28,029 [INFO]: isatab.py(write_study_table_files:1337) >> Writing 3 rows\n",
      "2021-01-20 21:50:28,043 [INFO]: isatab.py(_all_end_to_end_paths:1152) >> Found 3 paths!\n",
      "2021-01-20 21:50:28,065 [INFO]: isatab.py(_all_end_to_end_paths:1152) >> Found 3 paths!\n"
     ]
    }
   ],
   "source": [
    "dataframes = dump_tables_to_dataframes(investigation)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 153,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'s_study.txt'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'a_assay_2.txt'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'a_assay_1.txt'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "for key in dataframes.keys():\n",
    "    display(key)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 154,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample Name</th>\n",
       "      <th>Protocol REF</th>\n",
       "      <th>Extract Name</th>\n",
       "      <th>Protocol REF.1</th>\n",
       "      <th>Assay Name</th>\n",
       "      <th>Raw Data File</th>\n",
       "      <th>Protocol REF.2</th>\n",
       "      <th>Normalization Name</th>\n",
       "      <th>Derived Data File</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>sample_material-0</td>\n",
       "      <td>nucleic acid extraction</td>\n",
       "      <td>extract-0</td>\n",
       "      <td>sequencing_2</td>\n",
       "      <td>assay-name-0</td>\n",
       "      <td>sequenced-data-0</td>\n",
       "      <td>analysis</td>\n",
       "      <td>dt-name-0</td>\n",
       "      <td>derived-data-0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>sample_material-1</td>\n",
       "      <td>nucleic acid extraction</td>\n",
       "      <td>extract-1</td>\n",
       "      <td>sequencing_2</td>\n",
       "      <td>assay-name-1</td>\n",
       "      <td>sequenced-data-1</td>\n",
       "      <td>analysis</td>\n",
       "      <td>dt-name-1</td>\n",
       "      <td>derived-data-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>sample_material-2</td>\n",
       "      <td>nucleic acid extraction</td>\n",
       "      <td>extract-2</td>\n",
       "      <td>sequencing_2</td>\n",
       "      <td>assay-name-2</td>\n",
       "      <td>sequenced-data-2</td>\n",
       "      <td>analysis</td>\n",
       "      <td>dt-name-2</td>\n",
       "      <td>derived-data-2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         Sample Name             Protocol REF Extract Name Protocol REF.1  \\\n",
       "0  sample_material-0  nucleic acid extraction    extract-0   sequencing_2   \n",
       "1  sample_material-1  nucleic acid extraction    extract-1   sequencing_2   \n",
       "2  sample_material-2  nucleic acid extraction    extract-2   sequencing_2   \n",
       "\n",
       "     Assay Name     Raw Data File Protocol REF.2 Normalization Name  \\\n",
       "0  assay-name-0  sequenced-data-0       analysis          dt-name-0   \n",
       "1  assay-name-1  sequenced-data-1       analysis          dt-name-1   \n",
       "2  assay-name-2  sequenced-data-2       analysis          dt-name-2   \n",
       "\n",
       "  Derived Data File  \n",
       "0    derived-data-0  \n",
       "1    derived-data-1  \n",
       "2    derived-data-2  "
      ]
     },
     "execution_count": 154,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataframes['a_assay_2.txt']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 155,
   "metadata": {},
   "outputs": [],
   "source": [
    "# dataframes['a_assay_1.txt']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 156,
   "metadata": {},
   "outputs": [],
   "source": [
    "# dataframes['a_assay_2.txt']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 157,
   "metadata": {},
   "outputs": [],
   "source": [
    "# dataframe = dump_tables_to_dataframes(investigation)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 158,
   "metadata": {},
   "outputs": [],
   "source": [
    "# print(dataframe)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 159,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2021-01-20 21:50:28,235 [INFO]: isatab.py(_all_end_to_end_paths:1152) >> Found 3 paths!\n",
      "2021-01-20 21:50:28,238 [INFO]: isatab.py(write_study_table_files:1330) >> Rendered 3 paths\n",
      "2021-01-20 21:50:28,240 [INFO]: isatab.py(write_study_table_files:1337) >> Writing 3 rows\n",
      "2021-01-20 21:50:28,252 [INFO]: isatab.py(_all_end_to_end_paths:1152) >> Found 3 paths!\n",
      "2021-01-20 21:50:28,271 [INFO]: isatab.py(_all_end_to_end_paths:1152) >> Found 3 paths!\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "isatools.model.Investigation(identifier='', filename='', title='', submission_date='', public_release_date='', ontology_source_references=[isatools.model.OntologySource(name='OBI', file='', version='', description='Ontology for Biomedical Investigations', comments=[])], publications=[], contacts=[], studies=[isatools.model.Study(filename='s_study.txt', identifier='s1', title='My ISA Study', description='Like with the Investigation, we could use the class constructor to set some default values, but have chosen to demonstrate in this example the use of instance variables to set initial values.', submission_date='2016-11-03', public_release_date='2016-11-03', contacts=[isatools.model.Person(last_name='Robertson', first_name='Alice', mid_initials='', email='', phone='', fax='', address='', affiliation='University of Life', roles=[isatools.model.OntologyAnnotation(term='submitter', term_source=None, term_accession='', comments=[])], comments=[])], design_descriptors=[isatools.model.OntologyAnnotation(term='intervention design', term_source=isatools.model.OntologySource(name='OBI', file='', version='', description='Ontology for Biomedical Investigations', comments=[]), term_accession='http://purl.obolibrary.org/obo/OBI_0000115', comments=[])], publications=[isatools.model.Publication(pubmed_id='12345678', doi='10.1038/sdata.2016.18', author_list='A. Robertson, B. Robertson', title='Experiments with Elephants', status=isatools.model.OntologyAnnotation(term='published', term_source=None, term_accession='', comments=[]), comments=[])], factors=[], protocols=[isatools.model.Protocol(name='sample collection', protocol_type=isatools.model.OntologyAnnotation(term='sample collection', term_source=None, term_accession='', comments=[]), uri='', version='', parameters=[], components=[], comments=[]), isatools.model.Protocol(name='extraction', protocol_type=isatools.model.OntologyAnnotation(term='material extraction', term_source=None, term_accession='', comments=[]), uri='', version='', parameters=[], components=[], comments=[]), isatools.model.Protocol(name='sequencing_1', protocol_type=isatools.model.OntologyAnnotation(term='nucleic acid sequencing', term_source=None, term_accession='', comments=[]), uri='', version='', parameters=[], components=[], comments=[]), isatools.model.Protocol(name='nucleic acid extraction', protocol_type=isatools.model.OntologyAnnotation(term='nucleic acid extraction', term_source=None, term_accession='', comments=[]), uri='', version='', parameters=[], components=[], comments=[]), isatools.model.Protocol(name='sequencing_2', protocol_type=isatools.model.OntologyAnnotation(term='nucleic acid sequencing', term_source=None, term_accession='', comments=[]), uri='', version='', parameters=[], components=[], comments=[]), isatools.model.Protocol(name='analysis', protocol_type=isatools.model.OntologyAnnotation(term='sequence analysis data transformation', term_source=None, term_accession='', comments=[]), uri='', version='', parameters=[], components=[], comments=[])], assays=[isatools.model.Assay(measurement_type=isatools.model.OntologyAnnotation(term='genome sequencing', term_source=None, term_accession='', comments=[]), technology_type=isatools.model.OntologyAnnotation(term='nucleotide sequencing', term_source=None, term_accession='', comments=[]), technology_platform='', filename='a_assay_1.txt', data_files=[isatools.model.DataFile(filename='sequenced-data-0', label='Raw Data File', generated_from=[], comments=[]), isatools.model.DataFile(filename='sequenced-data-1', label='Raw Data File', generated_from=[], comments=[]), isatools.model.DataFile(filename='sequenced-data-2', label='Raw Data File', generated_from=[], comments=[])], samples=[isatools.model.Sample(name='sample_material-0', characteristics=[isatools.model.Characteristic(category=isatools.model.OntologyAnnotation(term='Organism', term_source=None, term_accession='', comments=[]), value=isatools.model.OntologyAnnotation(term='Homo Sapiens', term_source=isatools.model.OntologySource(name='NCBITaxon', file='', version='', description='NCBI Taxonomy', comments=[]), term_accession='http://purl.bioontology.org/ontology/NCBITAXON/9606', comments=[]), unit=None, comments=[])], factor_values=[], derives_from=[isatools.model.Source(name='source_material', characteristics=[], comments=[])], comments=[]), isatools.model.Sample(name='sample_material-1', characteristics=[isatools.model.Characteristic(category=isatools.model.OntologyAnnotation(term='Organism', term_source=None, term_accession='', comments=[]), value=isatools.model.OntologyAnnotation(term='Homo Sapiens', term_source=isatools.model.OntologySource(name='NCBITaxon', file='', version='', description='NCBI Taxonomy', comments=[]), term_accession='http://purl.bioontology.org/ontology/NCBITAXON/9606', comments=[]), unit=None, comments=[])], factor_values=[], derives_from=[isatools.model.Source(name='source_material', characteristics=[], comments=[])], comments=[]), isatools.model.Sample(name='sample_material-2', characteristics=[isatools.model.Characteristic(category=isatools.model.OntologyAnnotation(term='Organism', term_source=None, term_accession='', comments=[]), value=isatools.model.OntologyAnnotation(term='Homo Sapiens', term_source=isatools.model.OntologySource(name='NCBITaxon', file='', version='', description='NCBI Taxonomy', comments=[]), term_accession='http://purl.bioontology.org/ontology/NCBITAXON/9606', comments=[]), unit=None, comments=[])], factor_values=[], derives_from=[isatools.model.Source(name='source_material', characteristics=[], comments=[])], comments=[])], process_sequence=[isatools.model.Process(id=\"\". name=\"None\", executes_protocol=Protocol(\n",
       "    name=extraction\n",
       "    protocol_type=material extraction\n",
       "    uri=\n",
       "    version=\n",
       "    parameters=0 ProtocolParameter objects\n",
       "    components=0 OntologyAnnotation objects\n",
       "    comments=0 Comment objects\n",
       "), date=\"None\", performer=\"None\", inputs=[isatools.model.Sample(name='sample_material-0', characteristics=[isatools.model.Characteristic(category=isatools.model.OntologyAnnotation(term='Organism', term_source=None, term_accession='', comments=[]), value=isatools.model.OntologyAnnotation(term='Homo Sapiens', term_source=isatools.model.OntologySource(name='NCBITaxon', file='', version='', description='NCBI Taxonomy', comments=[]), term_accession='http://purl.bioontology.org/ontology/NCBITAXON/9606', comments=[]), unit=None, comments=[])], factor_values=[], derives_from=[isatools.model.Source(name='source_material', characteristics=[], comments=[])], comments=[])], outputs=[<isatools.model.Material object at 0x126e0c0a0>]), isatools.model.Process(id=\"\". name=\"assay-name-0\", executes_protocol=Protocol(\n",
       "    name=sequencing_1\n",
       "    protocol_type=nucleic acid sequencing\n",
       "    uri=\n",
       "    version=\n",
       "    parameters=0 ProtocolParameter objects\n",
       "    components=0 OntologyAnnotation objects\n",
       "    comments=0 Comment objects\n",
       "), date=\"None\", performer=\"None\", inputs=[<isatools.model.Material object at 0x126e0c0a0>], outputs=[isatools.model.DataFile(filename='sequenced-data-0', label='Raw Data File', generated_from=[], comments=[])]), isatools.model.Process(id=\"\". name=\"None\", executes_protocol=Protocol(\n",
       "    name=extraction\n",
       "    protocol_type=material extraction\n",
       "    uri=\n",
       "    version=\n",
       "    parameters=0 ProtocolParameter objects\n",
       "    components=0 OntologyAnnotation objects\n",
       "    comments=0 Comment objects\n",
       "), date=\"None\", performer=\"None\", inputs=[isatools.model.Sample(name='sample_material-1', characteristics=[isatools.model.Characteristic(category=isatools.model.OntologyAnnotation(term='Organism', term_source=None, term_accession='', comments=[]), value=isatools.model.OntologyAnnotation(term='Homo Sapiens', term_source=isatools.model.OntologySource(name='NCBITaxon', file='', version='', description='NCBI Taxonomy', comments=[]), term_accession='http://purl.bioontology.org/ontology/NCBITAXON/9606', comments=[]), unit=None, comments=[])], factor_values=[], derives_from=[isatools.model.Source(name='source_material', characteristics=[], comments=[])], comments=[])], outputs=[<isatools.model.Material object at 0x1270b7e80>]), isatools.model.Process(id=\"\". name=\"assay-name-1\", executes_protocol=Protocol(\n",
       "    name=sequencing_1\n",
       "    protocol_type=nucleic acid sequencing\n",
       "    uri=\n",
       "    version=\n",
       "    parameters=0 ProtocolParameter objects\n",
       "    components=0 OntologyAnnotation objects\n",
       "    comments=0 Comment objects\n",
       "), date=\"None\", performer=\"None\", inputs=[<isatools.model.Material object at 0x1270b7e80>], outputs=[isatools.model.DataFile(filename='sequenced-data-1', label='Raw Data File', generated_from=[], comments=[])]), isatools.model.Process(id=\"\". name=\"None\", executes_protocol=Protocol(\n",
       "    name=extraction\n",
       "    protocol_type=material extraction\n",
       "    uri=\n",
       "    version=\n",
       "    parameters=0 ProtocolParameter objects\n",
       "    components=0 OntologyAnnotation objects\n",
       "    comments=0 Comment objects\n",
       "), date=\"None\", performer=\"None\", inputs=[isatools.model.Sample(name='sample_material-2', characteristics=[isatools.model.Characteristic(category=isatools.model.OntologyAnnotation(term='Organism', term_source=None, term_accession='', comments=[]), value=isatools.model.OntologyAnnotation(term='Homo Sapiens', term_source=isatools.model.OntologySource(name='NCBITaxon', file='', version='', description='NCBI Taxonomy', comments=[]), term_accession='http://purl.bioontology.org/ontology/NCBITAXON/9606', comments=[]), unit=None, comments=[])], factor_values=[], derives_from=[isatools.model.Source(name='source_material', characteristics=[], comments=[])], comments=[])], outputs=[<isatools.model.Material object at 0x1270b7d60>]), isatools.model.Process(id=\"\". name=\"assay-name-2\", executes_protocol=Protocol(\n",
       "    name=sequencing_1\n",
       "    protocol_type=nucleic acid sequencing\n",
       "    uri=\n",
       "    version=\n",
       "    parameters=0 ProtocolParameter objects\n",
       "    components=0 OntologyAnnotation objects\n",
       "    comments=0 Comment objects\n",
       "), date=\"None\", performer=\"None\", inputs=[<isatools.model.Material object at 0x1270b7d60>], outputs=[isatools.model.DataFile(filename='sequenced-data-2', label='Raw Data File', generated_from=[], comments=[])])], other_material=[<isatools.model.Material object at 0x126e0c0a0>, <isatools.model.Material object at 0x1270b7e80>, <isatools.model.Material object at 0x1270b7d60>], characteristic_categories=[], comments=[], units=[]), isatools.model.Assay(measurement_type=isatools.model.OntologyAnnotation(term='transcription profiling', term_source=None, term_accession='', comments=[]), technology_type=isatools.model.OntologyAnnotation(term='nucleotide sequencing', term_source=None, term_accession='', comments=[]), technology_platform='', filename='a_assay_2.txt', data_files=[isatools.model.DataFile(filename='sequenced-data-0', label='Raw Data File', generated_from=[], comments=[]), isatools.model.DataFile(filename='sequenced-data-1', label='Raw Data File', generated_from=[], comments=[]), isatools.model.DataFile(filename='sequenced-data-2', label='Raw Data File', generated_from=[], comments=[])], samples=[isatools.model.Sample(name='sample_material-0', characteristics=[isatools.model.Characteristic(category=isatools.model.OntologyAnnotation(term='Organism', term_source=None, term_accession='', comments=[]), value=isatools.model.OntologyAnnotation(term='Homo Sapiens', term_source=isatools.model.OntologySource(name='NCBITaxon', file='', version='', description='NCBI Taxonomy', comments=[]), term_accession='http://purl.bioontology.org/ontology/NCBITAXON/9606', comments=[]), unit=None, comments=[])], factor_values=[], derives_from=[isatools.model.Source(name='source_material', characteristics=[], comments=[])], comments=[]), isatools.model.Sample(name='sample_material-1', characteristics=[isatools.model.Characteristic(category=isatools.model.OntologyAnnotation(term='Organism', term_source=None, term_accession='', comments=[]), value=isatools.model.OntologyAnnotation(term='Homo Sapiens', term_source=isatools.model.OntologySource(name='NCBITaxon', file='', version='', description='NCBI Taxonomy', comments=[]), term_accession='http://purl.bioontology.org/ontology/NCBITAXON/9606', comments=[]), unit=None, comments=[])], factor_values=[], derives_from=[isatools.model.Source(name='source_material', characteristics=[], comments=[])], comments=[]), isatools.model.Sample(name='sample_material-2', characteristics=[isatools.model.Characteristic(category=isatools.model.OntologyAnnotation(term='Organism', term_source=None, term_accession='', comments=[]), value=isatools.model.OntologyAnnotation(term='Homo Sapiens', term_source=isatools.model.OntologySource(name='NCBITaxon', file='', version='', description='NCBI Taxonomy', comments=[]), term_accession='http://purl.bioontology.org/ontology/NCBITAXON/9606', comments=[]), unit=None, comments=[])], factor_values=[], derives_from=[isatools.model.Source(name='source_material', characteristics=[], comments=[])], comments=[])], process_sequence=[isatools.model.Process(id=\"\". name=\"None\", executes_protocol=Protocol(\n",
       "    name=nucleic acid extraction\n",
       "    protocol_type=nucleic acid extraction\n",
       "    uri=\n",
       "    version=\n",
       "    parameters=0 ProtocolParameter objects\n",
       "    components=0 OntologyAnnotation objects\n",
       "    comments=0 Comment objects\n",
       "), date=\"None\", performer=\"None\", inputs=[isatools.model.Sample(name='sample_material-0', characteristics=[isatools.model.Characteristic(category=isatools.model.OntologyAnnotation(term='Organism', term_source=None, term_accession='', comments=[]), value=isatools.model.OntologyAnnotation(term='Homo Sapiens', term_source=isatools.model.OntologySource(name='NCBITaxon', file='', version='', description='NCBI Taxonomy', comments=[]), term_accession='http://purl.bioontology.org/ontology/NCBITAXON/9606', comments=[]), unit=None, comments=[])], factor_values=[], derives_from=[isatools.model.Source(name='source_material', characteristics=[], comments=[])], comments=[])], outputs=[<isatools.model.Material object at 0x12706fb50>]), isatools.model.Process(id=\"\". name=\"assay-name-0\", executes_protocol=Protocol(\n",
       "    name=sequencing_2\n",
       "    protocol_type=nucleic acid sequencing\n",
       "    uri=\n",
       "    version=\n",
       "    parameters=0 ProtocolParameter objects\n",
       "    components=0 OntologyAnnotation objects\n",
       "    comments=0 Comment objects\n",
       "), date=\"None\", performer=\"None\", inputs=[<isatools.model.Material object at 0x12706fb50>], outputs=[isatools.model.DataFile(filename='sequenced-data-0', label='Raw Data File', generated_from=[], comments=[])]), isatools.model.Process(id=\"\". name=\"dt-name-0\", executes_protocol=Protocol(\n",
       "    name=analysis\n",
       "    protocol_type=sequence analysis data transformation\n",
       "    uri=\n",
       "    version=\n",
       "    parameters=0 ProtocolParameter objects\n",
       "    components=0 OntologyAnnotation objects\n",
       "    comments=0 Comment objects\n",
       "), date=\"None\", performer=\"None\", inputs=[isatools.model.DataFile(filename='sequenced-data-0', label='Raw Data File', generated_from=[], comments=[])], outputs=[isatools.model.DataFile(filename='derived-data-0', label='Derived Data File', generated_from=[], comments=[])]), isatools.model.Process(id=\"\". name=\"None\", executes_protocol=Protocol(\n",
       "    name=nucleic acid extraction\n",
       "    protocol_type=nucleic acid extraction\n",
       "    uri=\n",
       "    version=\n",
       "    parameters=0 ProtocolParameter objects\n",
       "    components=0 OntologyAnnotation objects\n",
       "    comments=0 Comment objects\n",
       "), date=\"None\", performer=\"None\", inputs=[isatools.model.Sample(name='sample_material-1', characteristics=[isatools.model.Characteristic(category=isatools.model.OntologyAnnotation(term='Organism', term_source=None, term_accession='', comments=[]), value=isatools.model.OntologyAnnotation(term='Homo Sapiens', term_source=isatools.model.OntologySource(name='NCBITaxon', file='', version='', description='NCBI Taxonomy', comments=[]), term_accession='http://purl.bioontology.org/ontology/NCBITAXON/9606', comments=[]), unit=None, comments=[])], factor_values=[], derives_from=[isatools.model.Source(name='source_material', characteristics=[], comments=[])], comments=[])], outputs=[<isatools.model.Material object at 0x12706ff10>]), isatools.model.Process(id=\"\". name=\"assay-name-1\", executes_protocol=Protocol(\n",
       "    name=sequencing_2\n",
       "    protocol_type=nucleic acid sequencing\n",
       "    uri=\n",
       "    version=\n",
       "    parameters=0 ProtocolParameter objects\n",
       "    components=0 OntologyAnnotation objects\n",
       "    comments=0 Comment objects\n",
       "), date=\"None\", performer=\"None\", inputs=[<isatools.model.Material object at 0x12706ff10>], outputs=[isatools.model.DataFile(filename='sequenced-data-1', label='Raw Data File', generated_from=[], comments=[])]), isatools.model.Process(id=\"\". name=\"dt-name-1\", executes_protocol=Protocol(\n",
       "    name=analysis\n",
       "    protocol_type=sequence analysis data transformation\n",
       "    uri=\n",
       "    version=\n",
       "    parameters=0 ProtocolParameter objects\n",
       "    components=0 OntologyAnnotation objects\n",
       "    comments=0 Comment objects\n",
       "), date=\"None\", performer=\"None\", inputs=[isatools.model.DataFile(filename='sequenced-data-1', label='Raw Data File', generated_from=[], comments=[])], outputs=[isatools.model.DataFile(filename='derived-data-1', label='Derived Data File', generated_from=[], comments=[])]), isatools.model.Process(id=\"\". name=\"None\", executes_protocol=Protocol(\n",
       "    name=nucleic acid extraction\n",
       "    protocol_type=nucleic acid extraction\n",
       "    uri=\n",
       "    version=\n",
       "    parameters=0 ProtocolParameter objects\n",
       "    components=0 OntologyAnnotation objects\n",
       "    comments=0 Comment objects\n",
       "), date=\"None\", performer=\"None\", inputs=[isatools.model.Sample(name='sample_material-2', characteristics=[isatools.model.Characteristic(category=isatools.model.OntologyAnnotation(term='Organism', term_source=None, term_accession='', comments=[]), value=isatools.model.OntologyAnnotation(term='Homo Sapiens', term_source=isatools.model.OntologySource(name='NCBITaxon', file='', version='', description='NCBI Taxonomy', comments=[]), term_accession='http://purl.bioontology.org/ontology/NCBITAXON/9606', comments=[]), unit=None, comments=[])], factor_values=[], derives_from=[isatools.model.Source(name='source_material', characteristics=[], comments=[])], comments=[])], outputs=[<isatools.model.Material object at 0x1270d9730>]), isatools.model.Process(id=\"\". name=\"assay-name-2\", executes_protocol=Protocol(\n",
       "    name=sequencing_2\n",
       "    protocol_type=nucleic acid sequencing\n",
       "    uri=\n",
       "    version=\n",
       "    parameters=0 ProtocolParameter objects\n",
       "    components=0 OntologyAnnotation objects\n",
       "    comments=0 Comment objects\n",
       "), date=\"None\", performer=\"None\", inputs=[<isatools.model.Material object at 0x1270d9730>], outputs=[isatools.model.DataFile(filename='sequenced-data-2', label='Raw Data File', generated_from=[], comments=[])]), isatools.model.Process(id=\"\". name=\"dt-name-2\", executes_protocol=Protocol(\n",
       "    name=analysis\n",
       "    protocol_type=sequence analysis data transformation\n",
       "    uri=\n",
       "    version=\n",
       "    parameters=0 ProtocolParameter objects\n",
       "    components=0 OntologyAnnotation objects\n",
       "    comments=0 Comment objects\n",
       "), date=\"None\", performer=\"None\", inputs=[isatools.model.DataFile(filename='sequenced-data-2', label='Raw Data File', generated_from=[], comments=[])], outputs=[isatools.model.DataFile(filename='derived-data-2', label='Derived Data File', generated_from=[], comments=[])])], other_material=[<isatools.model.Material object at 0x12706fb50>, <isatools.model.Material object at 0x12706ff10>, <isatools.model.Material object at 0x1270d9730>], characteristic_categories=[], comments=[], units=[])], sources=[isatools.model.Source(name='source_material', characteristics=[], comments=[])], samples=[isatools.model.Sample(name='sample_material-0', characteristics=[isatools.model.Characteristic(category=isatools.model.OntologyAnnotation(term='Organism', term_source=None, term_accession='', comments=[]), value=isatools.model.OntologyAnnotation(term='Homo Sapiens', term_source=isatools.model.OntologySource(name='NCBITaxon', file='', version='', description='NCBI Taxonomy', comments=[]), term_accession='http://purl.bioontology.org/ontology/NCBITAXON/9606', comments=[]), unit=None, comments=[])], factor_values=[], derives_from=[isatools.model.Source(name='source_material', characteristics=[], comments=[])], comments=[]), isatools.model.Sample(name='sample_material-1', characteristics=[isatools.model.Characteristic(category=isatools.model.OntologyAnnotation(term='Organism', term_source=None, term_accession='', comments=[]), value=isatools.model.OntologyAnnotation(term='Homo Sapiens', term_source=isatools.model.OntologySource(name='NCBITaxon', file='', version='', description='NCBI Taxonomy', comments=[]), term_accession='http://purl.bioontology.org/ontology/NCBITAXON/9606', comments=[]), unit=None, comments=[])], factor_values=[], derives_from=[isatools.model.Source(name='source_material', characteristics=[], comments=[])], comments=[]), isatools.model.Sample(name='sample_material-2', characteristics=[isatools.model.Characteristic(category=isatools.model.OntologyAnnotation(term='Organism', term_source=None, term_accession='', comments=[]), value=isatools.model.OntologyAnnotation(term='Homo Sapiens', term_source=isatools.model.OntologySource(name='NCBITaxon', file='', version='', description='NCBI Taxonomy', comments=[]), term_accession='http://purl.bioontology.org/ontology/NCBITAXON/9606', comments=[]), unit=None, comments=[])], factor_values=[], derives_from=[isatools.model.Source(name='source_material', characteristics=[], comments=[])], comments=[])], process_sequence=[isatools.model.Process(id=\"\". name=\"None\", executes_protocol=Protocol(\n",
       "    name=sample collection\n",
       "    protocol_type=sample collection\n",
       "    uri=\n",
       "    version=\n",
       "    parameters=0 ProtocolParameter objects\n",
       "    components=0 OntologyAnnotation objects\n",
       "    comments=0 Comment objects\n",
       "), date=\"None\", performer=\"None\", inputs=[isatools.model.Source(name='source_material', characteristics=[], comments=[])], outputs=[isatools.model.Sample(name='sample_material-0', characteristics=[isatools.model.Characteristic(category=isatools.model.OntologyAnnotation(term='Organism', term_source=None, term_accession='', comments=[]), value=isatools.model.OntologyAnnotation(term='Homo Sapiens', term_source=isatools.model.OntologySource(name='NCBITaxon', file='', version='', description='NCBI Taxonomy', comments=[]), term_accession='http://purl.bioontology.org/ontology/NCBITAXON/9606', comments=[]), unit=None, comments=[])], factor_values=[], derives_from=[isatools.model.Source(name='source_material', characteristics=[], comments=[])], comments=[]), isatools.model.Sample(name='sample_material-1', characteristics=[isatools.model.Characteristic(category=isatools.model.OntologyAnnotation(term='Organism', term_source=None, term_accession='', comments=[]), value=isatools.model.OntologyAnnotation(term='Homo Sapiens', term_source=isatools.model.OntologySource(name='NCBITaxon', file='', version='', description='NCBI Taxonomy', comments=[]), term_accession='http://purl.bioontology.org/ontology/NCBITAXON/9606', comments=[]), unit=None, comments=[])], factor_values=[], derives_from=[isatools.model.Source(name='source_material', characteristics=[], comments=[])], comments=[]), isatools.model.Sample(name='sample_material-2', characteristics=[isatools.model.Characteristic(category=isatools.model.OntologyAnnotation(term='Organism', term_source=None, term_accession='', comments=[]), value=isatools.model.OntologyAnnotation(term='Homo Sapiens', term_source=isatools.model.OntologySource(name='NCBITaxon', file='', version='', description='NCBI Taxonomy', comments=[]), term_accession='http://purl.bioontology.org/ontology/NCBITAXON/9606', comments=[]), unit=None, comments=[])], factor_values=[], derives_from=[isatools.model.Source(name='source_material', characteristics=[], comments=[])], comments=[])])], other_material=[], characteristic_categories=[], comments=[], units=[])], comments=[])"
      ]
     },
     "execution_count": 159,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "isatab.dump(investigation,'./notebook-output/') "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 160,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2021-01-20 21:50:28,293 [INFO]: isatab.py(validate:4202) >> Loading... ./notebook-output/i_investigation.txt\n",
      "2021-01-20 21:50:28,420 [INFO]: isatab.py(validate:4204) >> Running prechecks...\n",
      "2021-01-20 21:50:28,475 [WARNING]: isatab.py(check_doi:2172) >> (W) DOI 10.1038/sdata.2016.18 does not conform to DOI format\n",
      "2021-01-20 21:50:28,476 [INFO]: isatab.py(validate:4225) >> Finished prechecks...\n",
      "2021-01-20 21:50:28,476 [INFO]: isatab.py(validate:4226) >> Loading configurations found in /Users/philippe/Documents/git/isa-api2/isa-api/isatools/resources/config/xml\n",
      "2021-01-20 21:50:28,502 [INFO]: isatab.py(validate:4231) >> Using configurations found in /Users/philippe/Documents/git/isa-api2/isa-api/isatools/resources/config/xml\n",
      "2021-01-20 21:50:28,503 [INFO]: isatab.py(validate:4233) >> Checking investigation file against configuration...\n",
      "2021-01-20 21:50:28,506 [INFO]: isatab.py(validate:4236) >> Finished checking investigation file\n",
      "2021-01-20 21:50:28,507 [INFO]: isatab.py(validate:4255) >> Loading... s_study.txt\n",
      "2021-01-20 21:50:28,510 [INFO]: isatab.py(validate:4261) >> Validating s_study.txt against default study table configuration\n",
      "2021-01-20 21:50:28,511 [INFO]: isatab.py(validate:4263) >> Checking Factor Value presence...\n",
      "2021-01-20 21:50:28,512 [INFO]: isatab.py(validate:4266) >> Checking required fields...\n",
      "2021-01-20 21:50:28,512 [INFO]: isatab.py(validate:4269) >> Checking generic fields...\n",
      "2021-01-20 21:50:28,513 [INFO]: isatab.py(validate:4277) >> Checking unit fields...\n",
      "2021-01-20 21:50:28,514 [INFO]: isatab.py(validate:4284) >> Checking protocol fields...\n",
      "2021-01-20 21:50:28,515 [INFO]: isatab.py(validate:4294) >> Checking ontology fields...\n",
      "2021-01-20 21:50:28,516 [INFO]: isatab.py(validate:4304) >> Checking study group size...\n",
      "2021-01-20 21:50:28,516 [INFO]: isatab.py(validate:4308) >> Finished validation on s_study.txt\n",
      "2021-01-20 21:50:28,517 [INFO]: isatab.py(validate:4345) >> Loading... a_assay_1.txt\n",
      "2021-01-20 21:50:28,520 [INFO]: isatab.py(validate:4353) >> Validating a_assay_1.txt against assay table configuration (genome sequencing, nucleotide sequencing)...\n",
      "2021-01-20 21:50:28,521 [INFO]: isatab.py(validate:4359) >> Checking Factor Value presence...\n",
      "2021-01-20 21:50:28,521 [INFO]: isatab.py(validate:4363) >> Checking required fields...\n",
      "2021-01-20 21:50:28,522 [INFO]: isatab.py(validate:4366) >> Checking generic fields...\n",
      "2021-01-20 21:50:28,524 [INFO]: isatab.py(validate:4377) >> Checking unit fields...\n",
      "2021-01-20 21:50:28,524 [INFO]: isatab.py(validate:4387) >> Checking protocol fields...\n",
      "2021-01-20 21:50:28,525 [WARNING]: isatab.py(check_protocol_fields:4014) >> (W) Protocol(s) of type ['nucleic acid extraction'] defined in the ISA-configuration expected as a between 'Sample Name' and 'Extract Name' but has not been found, in the file 'a_assay_1.txt'\n",
      "2021-01-20 21:50:28,526 [WARNING]: isatab.py(check_protocol_fields:4014) >> (W) Protocol(s) of type ['library construction'] defined in the ISA-configuration expected as a between 'Extract Name' and 'Assay Name' but has not been found, in the file 'a_assay_1.txt'\n",
      "2021-01-20 21:50:28,527 [WARNING]: isatab.py(validate:4392) >> (W) There are some protocol inconsistencies in a_assay_1.txt against ('genome sequencing', 'nucleotide sequencing') configuration\n",
      "2021-01-20 21:50:28,528 [INFO]: isatab.py(validate:4399) >> Checking ontology fields...\n",
      "2021-01-20 21:50:28,529 [INFO]: isatab.py(validate:4412) >> Checking study group size...\n",
      "2021-01-20 21:50:28,529 [INFO]: isatab.py(validate:4416) >> Finished validation on a_assay_1.txt\n",
      "2021-01-20 21:50:28,530 [INFO]: isatab.py(validate:4422) >> Checking consistencies between study sample table and assay tables...\n",
      "2021-01-20 21:50:28,531 [INFO]: isatab.py(validate:4427) >> Finished checking study sample table against assay tables...\n",
      "2021-01-20 21:50:28,531 [INFO]: isatab.py(validate:4345) >> Loading... a_assay_2.txt\n",
      "2021-01-20 21:50:28,534 [INFO]: isatab.py(validate:4353) >> Validating a_assay_2.txt against assay table configuration (transcription profiling, nucleotide sequencing)...\n",
      "2021-01-20 21:50:28,535 [INFO]: isatab.py(validate:4359) >> Checking Factor Value presence...\n",
      "2021-01-20 21:50:28,535 [INFO]: isatab.py(validate:4363) >> Checking required fields...\n",
      "2021-01-20 21:50:28,536 [WARNING]: isatab.py(check_required_fields:3683) >> (W) Required field 'Parameter Value[library layout]' not found in the file 'a_assay_2.txt'\n",
      "2021-01-20 21:50:28,536 [INFO]: isatab.py(validate:4366) >> Checking generic fields...\n",
      "2021-01-20 21:50:28,539 [INFO]: isatab.py(validate:4377) >> Checking unit fields...\n",
      "2021-01-20 21:50:28,540 [INFO]: isatab.py(validate:4387) >> Checking protocol fields...\n",
      "2021-01-20 21:50:28,541 [WARNING]: isatab.py(check_protocol_fields:4014) >> (W) Protocol(s) of type ['library construction'] defined in the ISA-configuration expected as a between 'Extract Name' and 'Assay Name' but has not been found, in the file 'a_assay_2.txt'\n",
      "2021-01-20 21:50:28,541 [WARNING]: isatab.py(validate:4392) >> (W) There are some protocol inconsistencies in a_assay_2.txt against ('transcription profiling', 'nucleotide sequencing') configuration\n",
      "2021-01-20 21:50:28,542 [INFO]: isatab.py(validate:4399) >> Checking ontology fields...\n",
      "2021-01-20 21:50:28,542 [INFO]: isatab.py(validate:4412) >> Checking study group size...\n",
      "2021-01-20 21:50:28,543 [INFO]: isatab.py(validate:4416) >> Finished validation on a_assay_2.txt\n",
      "2021-01-20 21:50:28,543 [INFO]: isatab.py(validate:4422) >> Checking consistencies between study sample table and assay tables...\n",
      "2021-01-20 21:50:28,544 [INFO]: isatab.py(validate:4427) >> Finished checking study sample table against assay tables...\n",
      "2021-01-20 21:50:28,752 [INFO]: utils.py(detect_isatab_process_pooling:89) >> Checking s_study.txt\n",
      "2021-01-20 21:50:28,753 [INFO]: utils.py(detect_isatab_process_pooling:98) >> Checking a_assay_1.txt\n",
      "2021-01-20 21:50:28,757 [INFO]: utils.py(detect_graph_process_pooling:69) >> Possible process pooling detected on:  sequencing_1\n",
      "2021-01-20 21:50:28,758 [INFO]: utils.py(detect_graph_process_pooling:69) >> Possible process pooling detected on:  sequencing_1\n",
      "2021-01-20 21:50:28,759 [INFO]: utils.py(detect_graph_process_pooling:69) >> Possible process pooling detected on:  sequencing_1\n",
      "2021-01-20 21:50:28,759 [INFO]: utils.py(detect_isatab_process_pooling:98) >> Checking a_assay_2.txt\n",
      "2021-01-20 21:50:28,764 [INFO]: utils.py(detect_graph_process_pooling:69) >> Possible process pooling detected on:  sequencing_2\n",
      "2021-01-20 21:50:28,765 [INFO]: utils.py(detect_graph_process_pooling:69) >> Possible process pooling detected on:  sequencing_2\n",
      "2021-01-20 21:50:28,765 [INFO]: utils.py(detect_graph_process_pooling:69) >> Possible process pooling detected on:  sequencing_2\n",
      "2021-01-20 21:50:28,766 [INFO]: utils.py(detect_graph_process_pooling:69) >> Possible process pooling detected on:  analysis\n",
      "2021-01-20 21:50:28,767 [INFO]: utils.py(detect_graph_process_pooling:69) >> Possible process pooling detected on:  analysis\n",
      "2021-01-20 21:50:28,767 [INFO]: utils.py(detect_graph_process_pooling:69) >> Possible process pooling detected on:  analysis\n",
      "2021-01-20 21:50:28,768 [INFO]: isatab.py(validate:4440) >> Finished validation...\n"
     ]
    }
   ],
   "source": [
    "my_json_report = isatab.validate(open(os.path.join('./notebook-output/', 'i_investigation.txt')))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 161,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'errors': [], 'warnings': [{'message': 'DOI is not valid format', 'supplemental': 'Found 10.1038/sdata.2016.18 in DOI field', 'code': 3002}, {'message': 'Missing Protocol declaration', 'supplemental': \"Protocol(s) of type ['nucleic acid extraction'] defined in the ISA-configuration expected as a between 'Sample Name' and 'Extract Name' but has not been found, in the file 'a_assay_1.txt'\", 'code': 1007}, {'message': 'Missing Protocol declaration', 'supplemental': \"Protocol(s) of type ['library construction'] defined in the ISA-configuration expected as a between 'Extract Name' and 'Assay Name' but has not been found, in the file 'a_assay_1.txt'\", 'code': 1007}, {'message': 'A required column in assay table is not present', 'supplemental': \"Required field 'Parameter Value[library layout]' not found in the file 'a_assay_2.txt'\", 'code': 4010}, {'message': 'Missing Protocol declaration', 'supplemental': \"Protocol(s) of type ['library construction'] defined in the ISA-configuration expected as a between 'Extract Name' and 'Assay Name' but has not been found, in the file 'a_assay_2.txt'\", 'code': 1007}], 'info': [{'message': 'Found -1 study groups in s_study.txt', 'supplemental': 'Found -1 study groups in s_study.txt', 'code': 5001}, {'message': 'Found -1 study groups in a_assay_1.txt', 'supplemental': 'Found -1 study groups in a_assay_1.txt', 'code': 5001}, {'message': 'Found -1 study groups in a_assay_2.txt', 'supplemental': 'Found -1 study groups in a_assay_2.txt', 'code': 5001}], 'validation_finished': True}\n"
     ]
    }
   ],
   "source": [
    "print(my_json_report)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "isa-api-py38",
   "language": "python",
   "name": "isa-api-py38"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
